NGitLab.Tests/ProjectVariableClientTests.cs

```diff
@@ -52,4 +55,99 @@ public async Task Test_project_variables()
+        var projectVariableClient = context.Client.GetProjectVariableClient(project.Id);
+
+        // Create
+        var variable = projectVariableClient.Create(new Variable
+        {
+            Key = "My_Key",
+            Value = "My value",
+            Description = "Some important variable",
+            Protected = true,
+            Type = VariableType.Variable,
+            Masked = false,
+            Raw = false,
+            Scope = "test/*",
+        });
+
+        Assert.That(variable.Key, Is.EqualTo("My_Key"));
+        Assert.That(variable.Value, Is.EqualTo("My value"));
+
+        if (context.IsGitLabVersionInRange(VersionRange.Parse("[16.2,)"), out _))
+        {
+            Assert.That(variable.Description, Is.EqualTo("Some important variable"));
+        }
+
+        Assert.That(variable.Protected, Is.EqualTo(true));
+        Assert.That(variable.Type, Is.EqualTo(VariableType.Variable));
+        Assert.That(variable.Masked, Is.EqualTo(false));
+        Assert.That(variable.Raw, Is.EqualTo(false));
+        Assert.That(variable.Scope, Is.EqualTo("test/*"));
+
+        // Update
+        var newScope = "integration/*";
+        variable = projectVariableClient.Update(variable.Key, new Variable
+        {
+            Value = "My value edited",
+            Protected = false,
+            Scope = newScope,
+        },
+        variable.Scope);
+
+        Assert.That(variable.Key, Is.EqualTo("My_Key"));
+        Assert.That(variable.Value, Is.EqualTo("My value edited"));
+        Assert.That(variable.Protected, Is.EqualTo(false));
+        Assert.That(variable.Scope, Is.EqualTo(newScope));
+
+        // Delete
+        var ex = Assert.Throws<GitLabException>(() => projectVariableClient.Delete(variable.Key, "wrongScope"));
+        Assert.That(ex!.StatusCode == HttpStatusCode.NotFound);
+
+        projectVariableClient.Delete(variable.Key, newScope);
+
+        var variables = projectVariableClient.All.ToList();
+        Assert.That(variables, Is.Empty);
+
+        // All
+        projectVariableClient.Create(new Variable { Key = "Variable1", Value = "test", Scope = "test/*" });
+        projectVariableClient.Create(new Variable { Key = "Variable2", Value = "test", Scope = "integration" });
+        projectVariableClient.Create(new Variable { Key = "Variable3", Value = "test", Scope = "*" });
+        variables = projectVariableClient.All.ToList();
+        Assert.That(variables, Has.Count.EqualTo(3));
+    }
+
+    [Test]
+    [NGitLabRetry]
+    public void Test_matching_scoped_variables()
+    {
+        var variable = new Variable
+        {
+            Key = "My_Key",
+            Value = "My value",
+            Description = "Some important variable",
+            Protected = true,
+            Type = VariableType.Variable,
+            Masked = false,
+            Raw = false,
+            Scope = "test/*",
+        };
+
+        Assert.That(variable.IsMatchForEnvironment("test/"), Is.True);
```

Comment:
Seems that IsMatchForEnvironment is only used in this test.

Here you are more testing the logic of the extension method rather than the logic of GitLab.

I would favorise a test like you did above to validate the scopes

Reply:
I added another test as showcase of the functionality c168edc26e0c5ecdad387a067a98d81b5abc92f1

@Toa741 Would you please allow me to let the other testcase in as well?

Reply:
What I would do:
- I would keep the test you added (`Test_matching_scoped_variables`)
- I would put a different value on the 2 different variables (but keep the same key)
- Instead of checking on our side the environment
  - We can have a first check that, when no environment is specified, it should return a 409 (conflict)
  - When specifying an environment, we can check it's the correct value of it that is returned

For me, checking the resolution of the environment is out of scope (it's the inner responsibility of GitLab, not on the client to do so, even in the tests).
So would simply keep  the modified `Test_matching_scoped_variables`, then remove completely the `Test_standalone_matching_scoped_variables` one (and the need of IsMatchForEnvironment)